### PR TITLE
ATL scatter gather functionality added

### PR DESCRIPTION
**Pull request by chinmay-ghosh**

### CODE REVIEW

File: src/common/atomic_queue.cpp [hunk 1]
```diff
@@ -559,6 +689,355 @@ void *process_queue(void *arg) {
...
         }
+
+        case ATOMIC_SCATTER_INDEX:
+            // Atomic scatter - indexed
+            if (msgPointer->flag & ATOMIC_WRITE_COMPLETED) {
+                // last write completed. If buffer is still allocated,
+                // deallocate and  remove the entry
+                if (msgPointer->flag & ATOMIC_BUFFER_ALLOCATED) {
+                    try {
+                        allocator->deallocate(ATOMIC_REGION_ID,
```
Comment (mashoodk) [inline src/common/atomic_queue.cpp:700]: Can this deallocate error be ignored? We might have completed deallocated, But ATOMIC_BUFFER_ALLOCATED might be still set.

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: src/cis/fam_cis_rpc.proto [hunk 2]
```diff
@@ -230,3 +234,32 @@ message Fam_Atomic_Response {
...
+    uint32 gid = 12;
+}
+
+message Fam_Atomic_SG_Indexed_Request {
+    uint64 regionid = 1;
+    uint64 offset = 2;
+    uint64 key = 3;
+    uint64 nelements = 4;
+    bytes elementindex = 5;
+    uint64 indexsize = 6;
```
Comment (faizan-barmawer) [inline src/cis/fam_cis_rpc.proto:258]: Is this indexsize or elementsize?

File: src/common/atomic_queue.cpp [hunk 3]
```diff
@@ -250,16 +265,33 @@ bool atomicQueue::isQempty() {
     return ((lcqData->size == 0) ? true : false);
 }
 
+// Populate string of indexes to array
+uint64_t *parseIndex(char *inpStr, uint64_t nElements) {
```
[the remark this answers is not in the document]
Comment (faizan-barmawer) [reply src/common/atomic_queue.cpp:269]: Yes, It better to go with "repeated" rpc interface type to send across the element index array.

File: src/common/atomic_queue.cpp [hunk 4]
```diff
@@ -559,6 +689,355 @@ void *process_queue(void *arg) {
...
+                            msgPointer->flag |= ATOMIC_WRITE_IN_PROGRESS;
+                            try {
+                                openfam_persist(localPointerB, bufferSize);
+                                openfam_persist(msgPointer,
+                                                sizeof(msgPointer->flag));
+                            } catch (...) {
+                                retStatus = PERSISTERROR;
+                            }
+                            try {
+                                openfam_persist(localPointerB, bufferSize);
```
Comment (faizan-barmawer) [inline src/common/atomic_queue.cpp:747]: You can combine try-catch for all openfam_persist calls, unless you want to perform some cleanup operation for any of these call failures.

File: src/memory_service/fam_memory_service_direct.cpp [hunk 5]
```diff
@@ -446,7 +490,133 @@ void Fam_Memory_Service_Direct::put_atomic(uint64_t regionId,
...
+    InpMsg.nodeAddrSize = nodeAddrSize;
+    memcpy(&InpMsg.nodeAddr, nodeAddr, nodeAddrSize);
+    InpMsg.dstDataGdesc.regionId = regionId;
+    InpMsg.dstDataGdesc.offset = offset;
+    InpMsg.inElements = nElements;
+    InpMsg.indexSize = elementSize;
+    InpMsg.key = key;
+    InpMsg.flag |= ATOMIC_GATHER_INDEX;
+
+    int ret = atomicQ[qId].push(&InpMsg, elementIndex);
```
Comment (faizan-barmawer) [inline src/memory_service/fam_memory_service_direct.cpp:607]: "elementIndex" is a local string pointer, coming from rpc request parameter. 
In the "push" function, hope we are copying the content of this string and not saving the string pointer.

Looks like you are doing this operation at line# 177-184 atomic_queue.cpp

**Review by faizan-barmawer — CHANGES_REQUESTED** [review]

**Review by mashoodk — APPROVED** [review]

**Review by faizan-barmawer — APPROVED** [review]
Code looks good to merge.

Regards,
Faizan

**Pull request merged**